README.md
# WaPoNe_ItalianPackage

**WaPoNe_ItalianPackage** prepares and sets Magento for the Italian market.

In details:
  - Loads Italian regions (province)
  - Sets 'general/country/default' config parameter to 'IT'
  - Sets 'general/store_information/country_id' config parameter to 'IT'
  - Sets 'shipping/origin/country_id' config parameter to 'IT'
  - Sets 'tax/defaults/country' config parameter to 'IT'
  - Creates tax rate for Italy ('Italy IVA 22% (W1)')
  - Creates tax rule for Italy ('Tax Rule for Italy (W1)'), all customer tax class and product tax class combinations

## Compatibility

- Magento CE:
  - 2.3.x

<br /><br />


CHANGELOG.md
# Change Log

## [1.0.1] - 25/03/2020
### Added
- Italian regions update

## [1.0.0] - 01/03/2020
### Added
- Initial release


Setup/Patch/Data/AddDataForItaly.php
<?php
/**
 * WaPoNe
 *
 * @category   WaPoNe
 * @package    WaPoNe_ItalianPackage
 * @author     Michele Fantetti
 * @copyright  Copyright (c) 2020 WaPoNe (https://www.fantetti.net)
 * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

namespace WaPoNe\ItalianPackage\Setup\Patch\Data;

use Magento\Directory\Setup\DataInstaller;
use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Framework\Setup\Patch\DataPatchInterface;
use Magento\Directory\Setup\Patch\Data\InitializeDirectoryData;

/**
 * Add Regions for Italy.
 */
class AddDataForItaly implements DataPatchInterface
{
    /**
     * @var ModuleDataSetupInterface
     */
    private $moduleDataSetup;

    /**
     * @var \Magento\Directory\Setup\DataInstallerFactory
     */
    private $dataInstallerFactory;

    /**
     * @param ModuleDataSetupInterface $moduleDataSetup
     * @param \Magento\Directory\Setup\DataInstallerFactory $dataInstallerFactory
     */
    public function __construct(
        ModuleDataSetupInterface $moduleDataSetup,
        \Magento\Directory\Setup\DataInstallerFactory $dataInstallerFactory
    ) {
        $this->moduleDataSetup = $moduleDataSetup;
        $this->dataInstallerFactory = $dataInstallerFactory;
    }

    /**
     * @inheritdoc
     */
    public function apply()
    {
        /** @var DataInstaller $dataInstaller */
        $dataInstaller = $this->dataInstallerFactory->create();
        $dataInstaller->addCountryRegions(
            $this->moduleDataSetup->getConnection(),
            $this->getDataForItaly()
        );
    }

    /**
     * Italy states data.
     *
     * @return array
     */
    private function getDataForItaly()
    {
        return [
            ['IT', 'AG', 'Agrigento'],
            ['IT', 'AL', 'Alessandria'],
            ['IT', 'AN', 'Ancona'],
            ['IT', 'AO', 'Aosta'],
            ['IT', 'AQ', 'L\'Aquila'],
            ['IT', 'AR', 'Arezzo'],
            ['IT', 'AP', 'Ascoli-Piceno'],
            ['IT', 'AT', 'Asti'],
            ['IT', 'AV', 'Avellino'],
            ['IT', 'BA', 'Bari'],
            ['IT', 'BT', 'Barletta-Andria-Trani'],
            ['IT', 'BL', 'Belluno'],
            ['IT', 'BN', 'Benevento'],
            ['IT', 'BG', 'Bergamo'],
            ['IT', 'BI', 'Biella'],
            ['IT', 'BO', 'Bologna'],
            ['IT', 'BZ', 'Bolzano'],
            ['IT', 'BS', 'Brescia'],
            ['IT', 'BR', 'Brindisi'],
            ['IT', 'CA', 'Cagliari'],
            ['IT', 'CL', 'Caltanissetta'],
            ['IT', 'CB', 'Campobasso'],
            ['IT', 'CE', 'Caserta'],
            ['IT', 'CT', 'Catania'],
            ['IT', 'CZ', 'Catanzaro'],
            ['IT', 'CH', 'Chieti'],
            ['IT', 'CO', 'Como'],
            ['IT', 'CS', 'Cosenza'],
            ['IT', 'CR', 'Cremona'],
            ['IT', 'KR', 'Crotone'],
            ['IT', 'CN', 'Cuneo'],
            ['IT', 'EN', 'Enna'],
            ['IT', 'FM', 'Fermo'],
            ['IT', 'FE', 'Ferrara'],
            ['IT', 'FI', 'Firenze'],
            ['IT', 'FG', 'Foggia'],
            ['IT', 'FC', 'Forli-Cesena'],
            ['IT', 'FR', 'Frosinone'],
            ['IT', 'GE', 'Genova'],
            ['IT', 'GO', 'Gorizia'],
            ['IT', 'GR', 'Grosseto'],
            ['IT', 'IM', 'Imperia'],
            ['IT', 'IS', 'Isernia'],
            ['IT', 'SP', 'La-Spezia'],
            ['IT', 'LT', 'Latina'],
            ['IT', 'LE', 'Lecce'],
            ['IT', 'LC', 'Lecco'],
            ['IT', 'LI', 'Livorno'],
            ['IT', 'LO', 'Lodi'],
            ['IT', 'LU', 'Lucca'],
            ['IT', 'MC', 'Macerata'],
            ['IT', 'MN', 'Mantova'],
            ['IT', 'MS', 'Massa-Carrara'],
            ['IT', 'MT', 'Matera'],
            ['IT', 'ME', 'Messina'],
            ['IT', 'MI', 'Milano'],
            ['IT', 'MO', 'Modena'],
            ['IT', 'MB', 'Monza-Brianza'],
            ['IT', 'NA', 'Napoli'],
            ['IT', 'NO', 'Novara'],
            ['IT', 'NU', 'Nuoro'],
            ['IT', 'OR', 'Oristano'],
            ['IT', 'PD', 'Padova'],
            ['IT', 'PA', 'Palermo'],
            ['IT', 'PR', 'Parma'],
            ['IT', 'PV', 'Pavia'],
            ['IT', 'PG', 'Perugia'],
            ['IT', 'PU', 'Pesaro-Urbino'],
            ['IT', 'PE', 'Pescara'],
            ['IT', 'PC', 'Piacenza'],
            ['IT', 'PI', 'Pisa'],
            ['IT', 'PT', 'Pistoia'],
            ['IT', 'PN', 'Pordenone'],
            ['IT', 'PZ', 'Potenza'],
            ['IT', 'PO', 'Prato'],
            ['IT', 'RG', 'Ragusa'],
            ['IT', 'RA', 'Ravenna'],
            ['IT', 'RC', 'Reggio-Calabria'],
            ['IT', 'RE', 'Reggio-Emilia'],
            ['IT', 'RI', 'Rieti'],
            ['IT', 'RN', 'Rimini'],
            ['IT', 'RM', 'Roma'],
            ['IT', 'RO', 'Rovigo'],
            ['IT', 'SA', 'Salerno'],
            ['IT', 'SS', 'Sassari'],
            ['IT', 'SV', 'Savona'],
            ['IT', 'SI', 'Siena'],
            ['IT', 'SR', 'Siracusa'],
            ['IT', 'SO', 'Sondrio'],
            ['IT', 'SU', 'Sud Sardegna'],
            ['IT', 'TA', 'Taranto'],
            ['IT', 'TE', 'Teramo'],
            ['IT', 'TR', 'Terni'],
            ['IT', 'TO', 'Torino'],
            ['IT', 'TP', 'Trapani'],
            ['IT', 'TN', 'Trento'],
            ['IT', 'TV', 'Treviso'],
            ['IT', 'TS', 'Trieste'],
            ['IT', 'UD', 'Udine'],
            ['IT', 'VA', 'Varese'],
            ['IT', 'VE', 'Venezia'],
            ['IT', 'VB', 'Verbano-Cusio-Ossola'],
            ['IT', 'VC', 'Vercelli'],
            ['IT', 'VR', 'Verona'],
            ['IT', 'VV', 'Vibo-Valentia'],
            ['IT', 'VI', 'Vicenza'],
            ['IT', 'VT', 'Viterbo'],
        ];
    }

    /**
     * @inheritdoc
     */
    public static function getDependencies()
    {
        return [
            InitializeDirectoryData::class,
        ];
    }

    /**
     * @inheritdoc
     */
    public function getAliases()
    {
        return [];
    }
}


Setup/Patch/Data/ConfigValuesForItaly.php
<?php
/**
 * WaPoNe
 *
 * @category   WaPoNe
 * @package    WaPoNe_ItalianPackage
 * @author     Michele Fantetti
 * @copyright  Copyright (c) 2020 WaPoNe (https://www.fantetti.net)
 * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

namespace WaPoNe\ItalianPackage\Setup\Patch\Data;

use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Framework\Setup\Patch\DataPatchInterface;

/**
 * Add Config Values for Italy.
 */
class ConfigValuesForItaly implements DataPatchInterface
{
    const TABLE_CORE_CONFIG_DATA = 'core_config_data';
    const PATH_GENERAL_COUNTY_DEFAULT = 'general/country/default';
    const PATH_GENERAL_STOREINFORMATION_COUNTRYID = 'general/store_information/country_id';
    const PATH_SHIPPING_ORIGIN_COUNTRYID = 'shipping/origin/country_id';
    const PATH_TAX_DEFAULTS_COUNTRY = 'tax/defaults/country';


    /**
     * @var ModuleDataSetupInterface
     */
    private $moduleDataSetup;

    /**
     * @param ModuleDataSetupInterface $moduleDataSetup
     */
    public function __construct(
        ModuleDataSetupInterface $moduleDataSetup
    ) {
        $this->moduleDataSetup = $moduleDataSetup;
    }

    /**
     * @inheritdoc
     */
    public function apply()
    {
        // General Country Default
        $this->moduleDataSetup->getConnection()->update(
            $this->moduleDataSetup->getTable(self::TABLE_CORE_CONFIG_DATA),
            $this->getValueIT(),
            $this->getGeneralCountryDefault()
        );
        // General StoreInformation CountryId
        $this->moduleDataSetup->getConnection()->update(
            $this->moduleDataSetup->getTable(self::TABLE_CORE_CONFIG_DATA),
            $this->getValueIT(),
            $this->getGeneralStoreInformationCountryId()
        );
        // Shipping Origin CountryId
        $this->moduleDataSetup->getConnection()->update(
            $this->moduleDataSetup->getTable(self::TABLE_CORE_CONFIG_DATA),
            $this->getValueIT(),
            $this->getShippingOriginCountryId()
        );
        // Tax Defaults Country
        $this->moduleDataSetup->getConnection()->update(
            $this->moduleDataSetup->getTable(self::TABLE_CORE_CONFIG_DATA),
            $this->getValueIT(),
            $this->getTaxDefaultsCountry()
        );
    }

    /**
     * General Country Default
     *
     * @return array
     */
    private function getValueIT()
    {
        return [
            'value' => 'IT'
        ];
    }

    /**
     * General Country Default
     *
     * @return array
     */
    private function getGeneralCountryDefault()
    {
        return [
            'scope="default"',
            'scope_id=0',
            'path=?' => self::PATH_GENERAL_COUNTY_DEFAULT
        ];
    }

    /**
     * General StoreInformation CountryId
     *
     * @return array
     */
    private function getGeneralStoreInformationCountryId()
    {
        return [
            'scope="default"',
            'scope_id=0',
            'path=?' => self::PATH_GENERAL_STOREINFORMATION_COUNTRYID
        ];
    }

    /**
     * Shipping Origin CountryId
     *
     * @return array
     */
    private function getShippingOriginCountryId()
    {
        return [
            'scope="default"',
            'scope_id=0',
            'path=?' => self::PATH_SHIPPING_ORIGIN_COUNTRYID
        ];
    }

    /**
     * Tax Defaults Country
     *
     * @return array
     */
    private function getTaxDefaultsCountry()
    {
        return [
            'scope="default"',
            'scope_id=0',
            'path=?' => self::PATH_TAX_DEFAULTS_COUNTRY
        ];
    }

    /**
     * @inheritdoc
     */
    public static function getDependencies()
    {
        return [];
    }

    /**
     * @inheritdoc
     */
    public function getAliases()
    {
        return [];
    }
}


Setup/Patch/Data/AddTaxForItaly.php
<?php
/**
 * WaPoNe
 *
 * @category   WaPoNe
 * @package    WaPoNe_ItalianPackage
 * @author     Michele Fantetti
 * @copyright  Copyright (c) 2020 WaPoNe (https://www.fantetti.net)
 * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

namespace WaPoNe\ItalianPackage\Setup\Patch\Data;

use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Framework\Setup\Patch\DataPatchInterface;
use Magento\Tax\Model\TaxClass\Source\Customer as CustomerTaxClassSource;
use Magento\Tax\Model\TaxClass\Source\Product as ProductTaxClassSource;
use Magento\Tax\Api\TaxRuleRepositoryInterface;
use Magento\Tax\Api\Data\TaxRuleInterfaceFactory;
use Magento\Tax\Api\TaxRateRepositoryInterface;
use Magento\Tax\Api\Data\TaxRateInterfaceFactory;
use Magento\Tax\Model\Calculation\RateFactory;
use Magento\Framework\Api\SearchCriteriaBuilder;
use Magento\Framework\Api\FilterBuilder;
use Psr\Log\LoggerInterface;

/**
 * Add Tax for Italy.
 */
class AddTaxForItaly implements DataPatchInterface
{
    const TABLE_TAX_CALCULATION_RATE = 'tax_calculation_rate';
    const TAX_RATE_CODE = 'Italy IVA 22% (W1)';
    const TAX_RULE_CODE = 'Tax Rule for Italy (W1)';
    const ID_COLUMN_VALUE = 'value';

    /**
     * @var ModuleDataSetupInterface
     */
    private $moduleDataSetup;

    /**
     * @var CustomerTaxClassSource
     */
    private $customerTaxClassSource;

    /**
     * @var ProductTaxClassSource
     */
    private $productTaxClassSource;

    /**
     * @var TaxRuleRepositoryInterface
     */
    private $taxRuleRepository;

    /**
     * @var TaxRuleInterfaceFactory
     */
    private $ruleFactory;

    /**
     * @var TaxRateRepositoryInterface
     */
    protected $taxRateRepository;

    /**
     * @var TaxRateInterfaceFactory
     */
    protected $rateFactory;

    /**
     * @var RateFactory
     */
    private $taxRateFactory;

    /**
     * @var SearchCriteriaBuilder
     */
    private $criteriaBuilder;

    /**
     * @var FilterBuilder
     */
    private $filterBuilder;

    /**
     * @var LoggerInterface
     */
    private $logger;

    /**
     * @param ModuleDataSetupInterface $moduleDataSetup
     * @param CustomerTaxClassSource $customerTaxClassSource
     * @param ProductTaxClassSource $productTaxClassSource
     * @param TaxRuleRepositoryInterface $taxRuleRepository
     * @param TaxRuleInterfaceFactory $ruleFactory
     * @param TaxRateRepositoryInterface $taxRateRepository
     * @param TaxRateInterfaceFactory $rateFactory
     * @param RateFactory $taxRateFactory
     * @param SearchCriteriaBuilder $criteriaBuilder
     * @param FilterBuilder $filterBuilder
     * @param LoggerInterface $logger
     */
    public function __construct(
        ModuleDataSetupInterface $moduleDataSetup,
        CustomerTaxClassSource $customerTaxClassSource,
        ProductTaxClassSource $productTaxClassSource,
        TaxRuleRepositoryInterface $taxRuleRepository,
        TaxRuleInterfaceFactory $ruleFactory,
        TaxRateRepositoryInterface $taxRateRepository,
        TaxRateInterfaceFactory $rateFactory,
        RateFactory $taxRateFactory,
        SearchCriteriaBuilder $criteriaBuilder,
        FilterBuilder $filterBuilder,
        LoggerInterface $logger
    ) {
        $this->moduleDataSetup = $moduleDataSetup;
        $this->customerTaxClassSource = $customerTaxClassSource;
        $this->productTaxClassSource = $productTaxClassSource;
        $this->taxRuleRepository = $taxRuleRepository;
        $this->ruleFactory = $ruleFactory;
        $this->taxRateRepository = $taxRateRepository;
        $this->rateFactory = $rateFactory;
        $this->taxRateFactory = $taxRateFactory;
        $this->criteriaBuilder = $criteriaBuilder;
        $this->filterBuilder = $filterBuilder;
        $this->logger = $logger;
    }

    /**
     * @inheritdoc
     */
    public function apply()
    {
        try {
            $this->createTaxRatesForItaly();
            $this->createTaxRuleForItaly();
        }
        catch (\Exception $e) {
            $this->logger->error($e->getMessage());
        }
    }

    /**
     * Create Tax Rates for Italy.
     *
     * @throws \Magento\Framework\Exception\InputException
     */
    private function createTaxRatesForItaly()
    {
        $italianTaxCalculationRate = [
            'tax_country_id' => 'IT',
            'tax_region_id' => 0,
            'tax_postcode' => '*',
            'code' => self::TAX_RATE_CODE,
            'rate' => 22.00
        ];
        // creating tax Italian rate
        $taxRate = $this->rateFactory->create();
        $taxRate->setCode($italianTaxCalculationRate['code'])
            ->setTaxCountryId($italianTaxCalculationRate['tax_country_id'])
            ->setTaxRegionId($italianTaxCalculationRate['tax_region_id'])
            ->setTaxPostcode($italianTaxCalculationRate['tax_postcode'])
            ->setRate($italianTaxCalculationRate['rate']);
        // saving tax Italian rate
        $this->taxRateRepository->save($taxRate);
    }

    /**
     * Create Tax Rule for Italy.
     *
     * @return void
     * @throws \Exception if something went wrong while saving the tax rule.
     */
    private function createTaxRuleForItaly()
    {
        $taxRuleForItaly = [
            'code'                  => self::TAX_RULE_CODE,
            'tax_rate'              => '',
            'tax_customer_class'    => $this->getCustomerTaxClassIds(),
            'tax_product_class'     => $this->getProductTaxClassIds(),
            'priority'              => 0,
            'calculate_subtotal'    => '',
            'position'              => 0
        ];

        // checking if the Tax Rule for Italy already exists
        $filter = $this->filterBuilder->setField('code')
            ->setConditionType('=')
            ->setValue($taxRuleForItaly['code'])
            ->create();
        $criteria = $this->criteriaBuilder->addFilters([$filter])->create();
        $existingRates = $this->taxRuleRepository->getList($criteria)->getItems();
        if (!empty($existingRates)) {
            throw new \Exception(self::TAX_RATE_CODE . ' already exists!');
        }

        // loading tax Italian rate
        $taxRate = $this->taxRateFactory->create()->loadByCode(self::TAX_RATE_CODE);
        //creating tax Italian rule
        $taxRule = $this->ruleFactory->create();
        $taxRule->setCode($taxRuleForItaly['code'])
            ->setTaxRateIds([$taxRate->getId()])
            ->setCustomerTaxClassIds($taxRuleForItaly['tax_customer_class'])
            ->setProductTaxClassIds($taxRuleForItaly['tax_product_class'])
            ->setPriority($taxRuleForItaly['priority'])
            ->setCalculateSubtotal($taxRuleForItaly['calculate_subtotal'])
            ->setPosition($taxRuleForItaly['position']);
        // saving tax Italian rule
        $this->taxRuleRepository->save($taxRule);
    }

    /**
     * Get all customer tax class ids
     *
     * @return array
     * @throws \Magento\Framework\Exception\StateException
     */
    private function getCustomerTaxClassIds()
    {
        return array_column($this->customerTaxClassSource->getAllOptions(), self::ID_COLUMN_VALUE);
    }

    /**
     * Get all product tax class ids
     *
     * @return array
     */
    private function getProductTaxClassIds()
    {
        $productTaxClassIds = array_column(
            $this->productTaxClassSource->getAllOptions(),
            self::ID_COLUMN_VALUE
        );
        // $this->productTaxClassSource->getAllOptions() gets also 0 value; Magento core bug???
        if (($key = array_search(0, $productTaxClassIds)) !== false) {
            unset($productTaxClassIds[$key]);
        }
        return $productTaxClassIds;
    }

    /**
     * @inheritdoc
     */
    public static function getDependencies()
    {
        return [];
    }

    /**
     * @inheritdoc
     */
    public function getAliases()
    {
        return [];
    }
}
